File: unisync_backend/lib/core/device_connection.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:unisync_backend/core/device_provider.dart';
import 'package:unisync_backend/models/device_info/device_info.model.dart';
import 'package:unisync_backend/models/device_message/device_message.model.dart';
import 'package:unisync_backend/utils/extensions/map.ext.dart';
import 'package:unisync_backend/utils/extensions/string.ext.dart';
import 'package:unisync_backend/utils/extensions/uint8list.ext.dart';
import 'package:unisync_backend/utils/logger.dart';

abstract class ConnectionListener {
  void onMessage(DeviceMessage message);
}

class DeviceConnection {
  DeviceConnection(this.socket, this.inputStream, this.info) {
    _listenInputStream();
  }

  final SecureSocket socket;
  final Stream<Uint8List> inputStream;
  final DeviceInfo info;

  var _isConnected = true;

  late final StreamSubscription<Uint8List> _streamSubscription;

  ConnectionListener? messageListener;

  void _listenInputStream() {
    _streamSubscription = inputStream.listen(
      (event) {
        messageListener?.onMessage(DeviceMessage.fromJson(event.string.toMap()));
      },
      cancelOnError: true,
      onDone: disconnect,
    );
  }

  void send(DeviceMessage message) {
    if (_isConnected) {
      socket.writeln(message.toJson().toJsonString());
    }
  }

  Future<void> disconnect() async {
    _isConnected = false;
    await _streamSubscription.cancel();
    await socket.close();
    DeviceProvider.remove(info);
    infoLog('DeviceConnection: ${info.name} disconnected!');
  }
}

File: unisync_android/app/src/main/java/com/anhquan/unisync/core/device/dependencies/DeviceConnection.kt
package com.anhquan.unisync.core.device.dependencies

import com.anhquan.unisync.core.providers.DeviceProvider
import com.anhquan.unisync.models.DeviceInfo
import com.anhquan.unisync.models.DeviceMessage
import com.anhquan.unisync.utils.fromJson
import com.anhquan.unisync.utils.infoLog
import com.anhquan.unisync.utils.runSingle
import com.anhquan.unisync.utils.runTask
import com.anhquan.unisync.utils.toJson
import io.reactivex.rxjava3.disposables.Disposable
import java.io.BufferedReader
import java.io.BufferedWriter
import javax.net.ssl.SSLSocket

class DeviceConnection(private val socket: SSLSocket) {
    interface ConnectionListener {
        fun onConnection()

        fun onMessage(message: DeviceMessage)

        fun onDisconnection()
    }

    interface ConnectionEmitter {
        fun sendMessage(message: DeviceMessage)
    }

    private val reader: BufferedReader = socket.inputStream.bufferedReader()
    private var writer: BufferedWriter = socket.outputStream.bufferedWriter()

    lateinit var deviceInfo: DeviceInfo
    private var listener: ConnectionListener? = null

    private lateinit var inputStreamDisposable: Disposable
    private var isConnected = true

    init {
        listenInputStream()
    }

    fun addMessageListener(listener: ConnectionListener) {
        this.listener = listener
        this.listener!!.onConnection()
    }

    private fun listenInputStream() {
        inputStreamDisposable = runTask(
            task = {
                while (true) {
                    val message = reader.readLine()
                    if (message != null) {
                        it.onNext(message)
                    } else {
                        disconnect()
                        break
                    }
                }
            },
            onResult = {
                infoLog("${deviceInfo.name}@${deviceInfo.ip}: Message received:\n$it")
                fromJson(
                    it,
                    DeviceMessage::class.java
                )?.let { message ->
                    listener?.onMessage(message)
                }
            }
        )
    }

    fun send(message: DeviceMessage) {
        if (isConnected) {
            runSingle(
                callback = {
                    writer.apply {
                        write(toJson(message))
                        flush()
                    }
                    infoLog("${deviceInfo.name}@${deviceInfo.ip}: Message sent:\n${toJson(message)}")
                },
                onError = {
                    disconnect()
                }
            )
        }
    }

    private fun disconnect() {
        isConnected = false
        try {
            inputStreamDisposable.dispose()
            socket.close()
            listener?.onDisconnection()
            DeviceProvider.remove(deviceInfo)
        } catch (_: Exception) {
        }
    }
}